SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -451,6 +450,54 @@ void USpatialNetDriver::CreateAndInitializeCoreClasses()
 		PlayerSpawner = NewObject<USpatialPlayerSpawner>();
 		SnapshotManager = MakeUnique<SpatialSnapshotManager>();
 
+		CreateAndInitializeLoadBalancingClasses();
+
+		Dispatcher->Init(SpatialWorkerFlags);
+		Sender->Init(this, &TimerManager, Connection->GetEventTracer());
+		Receiver->Init(this, Connection->GetEventTracer());
+		GlobalStateManager->Init(this);
+		SnapshotManager->Init(Connection, GlobalStateManager);
+		PlayerSpawner->Init(this);
```

Comment:
Does this need to exist before we have the flag set that allows us to receive player spawn requests.

Reply:
There is a way to avoid that with the following diff:
```
diff --git a/SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp b/SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp
index f6193e0f3..dd07735b1 100644
--- a/SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp
+++ b/SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp
@@ -447,7 +447,6 @@ void USpatialNetDriver::CreateAndInitializeCoreClasses()
 		GlobalStateManager = GameInstance->GetGlobalStateManager();
 		check(GlobalStateManager != nullptr);
 
-		PlayerSpawner = NewObject<USpatialPlayerSpawner>();
 		SnapshotManager = MakeUnique<SpatialSnapshotManager>();
 
 		CreateAndInitializeLoadBalancingClasses();
@@ -457,9 +456,7 @@ void USpatialNetDriver::CreateAndInitializeCoreClasses()
 		Receiver->Init(this, Connection->GetEventTracer());
 		GlobalStateManager->Init(this);
 		SnapshotManager->Init(Connection, GlobalStateManager);
-		PlayerSpawner->Init(this);
-		PlayerSpawner->OnPlayerSpawnFailed.BindUObject(GameInstance, &USpatialGameInstance::HandleOnPlayerSpawnFailed);
-
+		
 		SpatialMetrics->ControllerRefProvider.BindUObject(this, &USpatialNetDriver::GetCurrentPlayerControllerRef);
 
 		// PackageMap value has been set earlier in USpatialNetConnection::InitBase
@@ -510,6 +507,10 @@ void USpatialNetDriver::CreateAndInitializeCoreClassesAfterStartup()
 			InitialOnlyFilter = MakeUnique<SpatialGDK::InitialOnlyFilter>(*Connection);
 		}
 
+		PlayerSpawner = NewObject<USpatialPlayerSpawner>();
+		PlayerSpawner->Init(this);
+		PlayerSpawner->OnPlayerSpawnFailed.BindUObject(GameInstance, &USpatialGameInstance::HandleOnPlayerSpawnFailed);
+
 		const SpatialGDK::FSubView& ActorSubview = SpatialGDK::ActorSubviews::CreateActorSubView(*this);
 
 		const SpatialGDK::FSubView& ActorAuthSubview = SpatialGDK::ActorSubviews::CreateActorAuthSubView(*this);
@@ -894,7 +895,10 @@ void USpatialNetDriver::OnMapLoaded(UWorld* LoadedWorld)
 	{
 		if (ClientCanSendPlayerSpawnRequests())
 		{
-			MakePlayerSpawnRequest();
+			if (bIsReadyToStart)
+			{
+				MakePlayerSpawnRequest();
+			}
 		}
 		else
 		{
@@ -3261,6 +3265,8 @@ void USpatialNetDriver::TryFinishStartup()
 			Connection->SetStartupComplete();
 
 			CreateAndInitializeCoreClassesAfterStartup();
+
+			MakePlayerSpawnRequest();
 		}
 		else
 		{

```

This once again highlights the issues with the startup flow, but the change itself would be nice.

Reply:
This is kind of what I meant. Althought I guess it's mostly moot as clients won't send player creation requests until things are ready

Reply:
Clarification: We need this on clients, they will try to send a PlayerSpawnRequest during OnMapLoaded, but their startup finishes independently during TickDispatch, and is blocked on a flag raised during OnMapLoaded, so these are necessarily ordered.